src/lib.rs

```diff
@@ -39,12 +44,222 @@ pub enum Error {
     /// The range that we try to insert into the interval tree is overlapping
     /// with another node from the tree.
     #[error("Addresses are overlapping.{0:?} intersects with existing {1:?}")]
-    Overlap(allocation_engine::Range, allocation_engine::Range),
+    Overlap(Range, Range),
     /// A node state can be changed just from Free to Allocated, other transitions
     /// are not valid.
     #[error("Invalid state transition for node {0:?} from {1:?} to NodeState::Free")]
-    InvalidStateTransition(allocation_engine::Range, allocation_engine::NodeState),
+    InvalidStateTransition(Range, NodeState),
+    /// Address is unaligned
+    #[error("The address is not aligned.")]
+    UnalignedAddress,
 }
 
 /// Wrapper over std::result::Result
 pub type Result<T> = result::Result<T, Error>;
+
+/// A closed interval range [start, end] used to describe a memory slot that
+/// will be assigned to a device by the VMM. This structure represents the key
+/// of the Node object in the interval tree implementation.
+#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Hash, Ord, Debug)]
+pub struct Range {
+    /// Lower boundary of the interval.
+    start: u64,
+    /// Upper boundary of the interval.
+    end: u64,
+}
+
+#[allow(clippy::len_without_is_empty)]
+impl Range {
+    /// Create a new Range object.
+    pub fn new(start: u64, end: u64) -> Result<Self> {
+        if start > end {
```

Comment:
Related to the comment about `len()`: what if `start == end`? Should this also be an error?

Reply:
In the light of https://github.com/rust-vmm/vm-allocator/issues/42 it actually makes sense to have a cell of length of exactly 1.